### PR TITLE
Custom encoding requires access to op-code peeking

### PR DESCRIPTION
**Pull request by dim**
Requires code constants to be public, but I don't see much harm in that as this implementation is unlikely to change. Added test to demonstrate the issue (e.g. detection and correct encoding/decoding of nil-values). Benchmark results are below:

master:

```
BenchmarkBool 10000000         118 ns/op         0 B/op        0 allocs/op
BenchmarkInt0 10000000         123 ns/op         0 B/op        0 allocs/op
BenchmarkInt1 10000000         167 ns/op         0 B/op        0 allocs/op
BenchmarkInt2  5000000         244 ns/op         0 B/op        0 allocs/op
BenchmarkInt4  5000000         246 ns/op         0 B/op        0 allocs/op
BenchmarkInt8  5000000         267 ns/op         0 B/op        0 allocs/op
BenchmarkIntBinary   3000000         440 ns/op        24 B/op        3 allocs/op
BenchmarkIntMsgpack2   2000000         793 ns/op         8 B/op        1 allocs/op
BenchmarkIntMsgpack3   2000000         699 ns/op         8 B/op        1 allocs/op
BenchmarkTime  3000000         464 ns/op         0 B/op        0 allocs/op
BenchmarkDuration  5000000         266 ns/op         0 B/op        0 allocs/op
BenchmarkBytes    500000        2573 ns/op      1024 B/op        1 allocs/op
BenchmarkMapStringString   1000000        1515 ns/op        16 B/op        4 allocs/op
BenchmarkMapStringStringPtr   300000        3996 ns/op       208 B/op       12 allocs/op
BenchmarkMapIntInt    300000        4847 ns/op       384 B/op       16 allocs/op
BenchmarkStringSlice   2000000         671 ns/op        10 B/op        2 allocs/op
BenchmarkStringSlicePtr  1000000        1564 ns/op        74 B/op        4 allocs/op
BenchmarkStruct   100000       18352 ns/op      3145 B/op       28 allocs/op
BenchmarkStructManual   200000       11246 ns/op      2720 B/op       21 allocs/op
BenchmarkStructMsgpack2    50000       25506 ns/op      3616 B/op       70 allocs/op
BenchmarkStructMsgpack3    50000       31869 ns/op      5550 B/op       37 allocs/op
BenchmarkStructJSON    10000      121426 ns/op      9336 B/op       30 allocs/op
BenchmarkStructGOB     10000      170774 ns/op     23925 B/op      450 allocs/op
BenchmarkCSV    100000       20378 ns/op      8764 B/op       13 allocs/op
BenchmarkCSVMsgpack   500000        2434 ns/op       384 B/op       13 allocs/op
```

custom-peeking:

```
BenchmarkBool 10000000         125 ns/op         0 B/op        0 allocs/op
BenchmarkInt0 10000000         139 ns/op         0 B/op        0 allocs/op
BenchmarkInt1 10000000         177 ns/op         0 B/op        0 allocs/op
BenchmarkInt2  5000000         274 ns/op         0 B/op        0 allocs/op
BenchmarkInt4  5000000         270 ns/op         0 B/op        0 allocs/op
BenchmarkInt8  5000000         291 ns/op         0 B/op        0 allocs/op
BenchmarkIntBinary   3000000         462 ns/op        24 B/op        3 allocs/op
BenchmarkIntMsgpack2   2000000         796 ns/op         8 B/op        1 allocs/op
BenchmarkIntMsgpack3   2000000         700 ns/op         8 B/op        1 allocs/op
BenchmarkTime  3000000         490 ns/op         0 B/op        0 allocs/op
BenchmarkDuration  5000000         276 ns/op         0 B/op        0 allocs/op
BenchmarkBytes    500000        2700 ns/op      1024 B/op        1 allocs/op
BenchmarkMapStringString   1000000        1614 ns/op        16 B/op        4 allocs/op
BenchmarkMapStringStringPtr   300000        4336 ns/op       208 B/op       12 allocs/op
BenchmarkMapIntInt    300000        4993 ns/op       384 B/op       16 allocs/op
BenchmarkStringSlice   2000000         750 ns/op        10 B/op        2 allocs/op
BenchmarkStringSlicePtr  1000000        1573 ns/op        74 B/op        4 allocs/op
BenchmarkStruct   100000       18735 ns/op      3143 B/op       28 allocs/op
BenchmarkStructManual   200000       11052 ns/op      2720 B/op       21 allocs/op
BenchmarkStructMsgpack2    50000       27786 ns/op      3616 B/op       70 allocs/op
BenchmarkStructMsgpack3    50000       31579 ns/op      5550 B/op       37 allocs/op
BenchmarkStructJSON    10000      124445 ns/op      9337 B/op       30 allocs/op
BenchmarkStructGOB     10000      184581 ns/op     23925 B/op      450 allocs/op
BenchmarkCSV    100000       19866 ns/op      8764 B/op       13 allocs/op
BenchmarkCSVMsgpack   500000        2634 ns/op       384 B/op       13 allocs/op
```

### CODE REVIEW

**Comment by dim** [comment]
Also, added a https://github.com/dim/msgpack/compare/vmihailenco:master...custom-peeking-no-cleanup branch, which reverts the clean-up methods. Less pretty, but a tad faster.

**Comment by vmihailenco** [comment]
If we have to export codes I would ask you to create a subpackage `msgpack/codes`. Then you will have `codes.Map16` and `codes.IsFixNum`. This way we don't clutter API.

File: custom_test.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+func decodeCustomTestSet(d *msgpack.Decoder, v reflect.Value) error {
+	if code, err := d.PeekCode(); err != nil {
+		return err
+	} else if code == msgpack.NilCode {
+		_, err = d.DecodeInterface()
+		return err
+	}
+
+	sln, err := d.DecodeSliceLen()
```
Comment (vmihailenco) [inline custom_test.go]: DecodeSliceLen should return `-1, nil` when slice is nil. So you can remove code above and replace it with:

```
	sln, err := d.DecodeSliceLen()
	if err != nil {
		return err
	}
	if sln == -1 {
		// Slice is nil
		return nil
	}
```

Does it work work for you?

**Comment by dim** [comment]
True, it would work for slices, but there are other use cases. I have pushed #25, which would cause only minimal disruption.

**Comment by vmihailenco** [comment]
The behaviour of decoding nil values was changed recently: https://github.com/vmihailenco/msgpack/commit/d659bf7478b5ddd2c6a72774029ebb2ed5ffb5ed . Previously this lib handled decoding nil values for all types.

Having that in mind I would prefer to restore old behaviour rather than requiring users to manually care about nil values. E.g. something like this https://github.com/vmihailenco/msgpack/pull/26, but for all types.

But first I would like to understand the issue we are fixing here. Can you please describe "other use cases" you have in mind? Is https://github.com/vmihailenco/msgpack/pull/26 expectable behaviour?

**Comment by dim** [comment]
Check out this one as an example:
https://github.com/vmihailenco/msgpack/pull/25/files#diff-3df1114f2fd22239fbcd3a8751375101R65

Assuming I have the following situation:
```
type Parent struct {
  T string
  C *Child
  F float64
}

type Child struct {
  N int
  S string
}

var parent *Parent
msgpack.NewDecoder(r).Decode(&parent)
```

Let's also assume, my `Parent` is packed over-efficiently to minimise space, so the order of attributes in my binary stream is `{T,C.N,C.S,F}`. It could be either something like `{string,int,string,float}` or `{string,nil,float}`, depending on Child being nil or not. To decode, I need to be able to peek or some sort of `IsNilNext` function. 

If I did what you suggest in #26, I would first decode `T`, then call `dec.DecodeInt()` to which could return `0, nil`, but I would not be able to tell if `C` itself is `nil` or if `C.N` is just `0`. How would I determine what to do next, e.g. should I call `DecodeString` to get `C.S` or `DecodeFloat` to get parent's `F`?

Makes sense?

**Comment by vmihailenco** [comment]
Makes sense. I don't mind accepting this PR if you:
- move codes to subpackage as was suggested (codes.MapCode and codes.IsFixNum).
- export `PeekCode() (byte, error)` but don't use it internally to not introduce unnecessary slowdown.

Sounds good?

**Comment by dim** [comment]
I really wouldn't mind keeping everything private and just expose a simple `DecodeNil` or `DecodeIfNil` as proposed in #25

**Comment by vmihailenco** [comment]
`DecodeIfNil` is a special case for `PeekCode`, so I would prefer to add `PeekCode.`

File: custom_test.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	"reflect"
+	"testing"
+
+	"gopkg.in/vmihailenco/msgpack.v2"
+)
+
+// ------------------------------------------------------------------------
+
+func init() {
+	msgpack.Register(reflect.TypeOf(customTestSet{}), encodeCustomTestSet, decodeCustomTestSet)
```
Comment (vmihailenco) [inline custom_test.go]: BTW please use `msgpack.Marshal`/`msgpack.Unmarshal`. I plan to remove `msgpack.Register` from the future version.

**Pull request closed without merging**